[Serve] Add `policy` config to `AutoscalingConfig`

GeneDer opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Move `BasicAutoscalingPolicy` to a public file and add the new `policy` config to `AutoscalingConfig`. 

## Related issue number

First PR for https://github.com/ray-project/ray/issues/41135

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

GeneDer commented on python/ray/serve/_private/utils.py:
```diff
@@ -644,3 +645,82 @@ def get_capacity_adjusted_num_replicas(
     return max(1, int(rounded_adjusted_num_replicas))
+
+
+def calculate_desired_num_replicas(
```
This is just moved form `autoscaling_policy.py`. No logical changes.

edoakes replied:
This really doesn't belong in `utils.py`, please either move it to `autoscaling_policy.py` and underscore prefix it so it's clearly not a public API or keep it in `_private/autoscaling_policy.py`

GeneDer commented:
@edoakes would be nice if you can give this a look as well

shrekris-anyscale commented on python/ray/serve/autoscaling_policy.py:
```diff
@@ -5,91 +5,17 @@
-    return desired_num_replicas
-
-
+@PublicAPI(stability="stable")
```
Should we mark these as `stable` now? `alpha`/`beta` seems safer until we get more users using custom policies.

GeneDer replied:
Yea, I was debating whether this should be stable or alpha/ beta and chose to mark it stable due to it's the current default, will mark it as beta instead

edoakes replied:
These should all be marked `alpha` for now

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -60,6 +64,9 @@ class AutoscalingConfig(BaseModel):
     # How long to wait before scaling up replicas
     upscale_delay_s: NonNegativeFloat = 30.0
 
+    # Custom autoscaling config. Defaulting to the request based autoscaler.
```
```suggestion
    # Custom autoscaling config. Defaulting to the request-based autoscaler.
```

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -22,6 +25,7 @@
 from ray.util.annotations import Deprecated, PublicAPI
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
+DEFAULT_AUTOSCALING_POLICY = "ray.serve.autoscaling_policy:BasicAutoscalingPolicy"
```
Could we move this to `constants.py`?

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
 
+    @validator("policy", always=True)
+    def serialize_policy(cls, policy, values):
+        """Serialize policy to be cloudpickled and base64 encoded utf8 string.
```
```suggestion
        """Serialize policy with cloudpickle into base64-encoded utf8 string.
```

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
+    def serialize_policy(cls, policy, values):
+        """Serialize policy to be cloudpickled and base64 encoded utf8 string.
+
+        If policy is a string, it is possible to be already serialized when constructed
+        from protobuf. In this case, we just return the already serialized string. Else
+        we import the policy and serialize it before return.
+        """
```
```suggestion
        If policy is a string, it's either (1) a serialized callable or (2) an import path. In the case of (1), we just return the already-serialized callable. For (2),
        we import the policy and serialize it before returning.
        """
```

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
+                imported_policy = import_attr(policy)
+            except AssertionError:
+                imported_policy = import_attr(policy)
+        else:
```
When do we expect this `else` case to happen?

GeneDer replied:
good call! let me remove this!

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
+            imported_policy = policy
+        elif isinstance(policy, str):
+            try:
+                assert b64encode(b64decode(policy)).decode("utf-8") == policy
```
Why can't we directly decode the policy with `utf-8`? E.g.:

```python
policy.decode("utf-8") == policy
```

GeneDer replied:
This is to ensure policy is already base64 https://stackoverflow.com/a/45928164

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
+        return b64encode(cloudpickle.dumps(imported_policy)).decode("utf-8")
+
+    def get_policy(self) -> Callable:
+        """Deserialize policy from cloudpicked base64 encoded utf8 string."""
```
```suggestion
        """Deserialize policy from cloudpickled, base64-encoded, utf8 string."""
```

edoakes replied:
why is this base64 encoded as a string?

GeneDer replied:
Because base64 encode will return a byte object and I don't want to have handle 3 data types as the input for policy 🙏

edoakes replied:
why not just use `bytes` directly?

GeneDer replied:
Because this config came from the users. The support for string and callable is required. Bytes are the optional one.

edoakes replied:
My understanding is that users can pass a callable or an import path. In both of these cases, we will resolve it to cloudpickled bytes of the actual class definition.

In that case, the type of this field should be `bytes` and the layer above can resolve from `Union[str, Callable]` into bytes when the config object is constructed.

Am I missing something?

GeneDer replied:
Yea, good idea! Can make a public `policy` for users to pass in the scaling function, and have a private `policy_def` or something that's in bytes.

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
+
+        return b64encode(cloudpickle.dumps(imported_policy)).decode("utf-8")
+
+    def get_policy(self) -> Callable:
```
Could you add a unit test for this where the custom policy isn't provided inside the Ray package? We should make sure this works nicely with runtime environments.

GeneDer replied:
The runtime env was only working during my test because I was using the same name for the custom scaling function. Will go back to the drawing board for it.

GeneDer replied:
Talk with Shreyas and Ed offline. For now we will make a hard requirement for the custom autoscaling policy to exist in the cluster env. We will add a follow up when if the user complains about using this with runtime env. 

And in that case we can refactor the Autoscaling Manager into a Ray actor running in runtime env and constantly update an instance variable for the deployment replica num which controller can fetch and update.

shrekris-anyscale commented on python/ray/serve/_private/utils.py:
```diff
@@ -644,3 +645,82 @@ def get_capacity_adjusted_num_replicas(
+    current_num_ongoing_requests: List[float],
+    override_min_replicas: Optional[float] = None,
+    override_max_replicas: Optional[float] = None,
+) -> int:  # (desired replicas):
```
```suggestion
) -> int:
```

shrekris-anyscale commented on src/ray/protobuf/serve.proto:
```diff
@@ -60,6 +60,9 @@ message AutoscalingConfig {
   // The multiplicative "gain" factor to limit downscale.
   optional double downscale_smoothing_factor = 11;
+
+  // The cloudpickle base64 serialized callable for the autoscaling policy.
```
Based on [this comment](https://github.com/ray-project/ray/pull/42072/files#diff-2f5ba63ace082b7b7cfeceb2baf9b2d9f805f7559ac81e1fa5878e7b33aa69b9R98-R101), it sounds like this can also be an import path. If so, could you mention that in this docstring? If not, could you change that comment?

```suggestion
  // The cloudpickled base64-serialized callable for the autoscaling policy.
```

shrekris-anyscale reviewed (commented):
Nice work! I left some comments.

edoakes commented on python/ray/serve/autoscaling_policy.py:
```diff
@@ -129,6 +55,7 @@ def get_decision_num_replicas(
         return curr_target_num_replicas
 
 
+@PublicAPI(stability="stable")
```
"Basic" is quite a bad name. It's unclear why it's "basic" and it doesn't describe how the policy works at all...

Let's rename it. I'm open to something like:

- `DefaultAutoscalingPolicy` (makes it clear that it's the default but doesn't tell you what it does)
- `ReplicaQueueLengthAutoscalingPolicy` (describes the behavior but doesn't tell you it's the default)

Or maybe we could have `DefaultAutoscalingPolicy` as an alias for `ReplicaQueueLengthAutoscalingPolicy`

edoakes commented on python/ray/serve/config.py:
```diff
@@ -84,6 +91,33 @@ def replicas_settings_valid(cls, max_replicas, values):
+    def serialize_policy(cls, policy, values):
+        """Serialize policy to be cloudpickled and base64 encoded utf8 string.
+
+        If policy is a string, it is possible to be already serialized when constructed
+        from protobuf. In this case, we just return the already serialized string. Else
+        we import the policy and serialize it before return.
```
Why do we need to base64 encode the string? It seems very unlikely to me that this is actually needed. If it is, it should be clearly documented why we need this extra level of encoding

And why do we need this dynamic/branching behavior? Instead we can just have the constructor always take in a serialized policy

GeneDer replied:
This is likely the same reason as https://github.com/ray-project/ray/blob/bbbe66200eafda8fd0be62aafff3ba40c071bb21/python/ray/serve/_private/config.py#L222. Somehow if we just pass in a raw string, then it's getting base64 encoded. But likely will not be needed after redesign to ensure runtime_env works.

edoakes commented:
Mostly concerned about this base64 encoding and dynamic behavior of the policy field.

If something nonstandard like this is required in a PR I expect it to be called out in the description to let reviewers know why it's necessary and what alternatives there are, if any.

GeneDer commented:
@edoakes @shrekris-anyscale I have addressed all the comments and removed the need to serialize policy with base64 (thanks to the idea on separating them into 2 fields). PTAL again when you have a sec!

edoakes commented on src/ray/protobuf/serve.proto:
```diff
@@ -60,6 +60,14 @@ message AutoscalingConfig {
   // The multiplicative "gain" factor to limit downscale.
   optional double downscale_smoothing_factor = 11;
+
+  // The policy can be passed in as a string import path or callable by the user. When
+  // the policy is passed in as a callable, the policy will be just the name of the
+  // function.
+  optional string policy = 12;
+
+  // The cloudpickled policy definition.
+  bytes serialized_policy_def = 13;
```
why do both need to be stored in the autoscaling config?

GeneDer replied:
oh good point, I think only `serialized_policy_def` is needed for deserializing later on. Let me drop policy

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +92,28 @@ def replicas_settings_valid(cls, max_replicas, values):
+    def serialize_policy(cls, policy, values):
+        """Serialize policy with cloudpickle.
+
+        If policy is a callable, cloudpickle the function and set on the
+        `serialized_policy_def` if not already. Also, return the name of the function as
+        the policy. If policy is a string, import the function, cloudpickle it, and set
+        on the `serialized_policy_def` if not already.
```
```suggestion
        If policy is a callable, cloudpickle the function and set the
        `serialized_policy_def` if it's empty. Also, return the name of the function as
        the policy. If policy is a string, import the function, cloudpickle it, and set
        the `serialized_policy_def` if it's empty.
```

shrekris-anyscale commented on python/ray/serve/config.py:
```diff
@@ -84,6 +92,28 @@ def replicas_settings_valid(cls, max_replicas, values):
         return max_replicas
 
+    @validator("policy", always=True)
+    def serialize_policy(cls, policy, values):
```
```suggestion
    def serialize_policy(cls, policy, values) -> str:
```

GeneDer commented on python/ray/serve/autoscaling_policy.py:
```diff
@@ -269,3 +272,13 @@ def get_current_lower_bound(
+
+
+@PublicAPI(stability="alpha")
+class DefaultAutoscalingPolicy(ReplicaQueueLengthAutoscalingPolicy):
```
I kinda just naturally do that XD will change to assignment

GeneDer replied:
oh actually if we use an assignment, then we can't annotate the api separately. (not that it matters in this case, but just found this difference)

GeneDer commented on python/ray/serve/_private/config.py:
```diff
@@ -199,6 +204,13 @@ def to_proto_bytes(self):
+        # according.
+        #
+        # See:
+        # https://github.com/protocolbuffers/protobuf/issues/2801#issuecomment-285812695
```
@shrekris-anyscale let me know if this comment helps to explain it a bit more or if we should add more content.

edoakes replied:
Alternatively we can just avoid using `MessageToDict`. The purpose of this utility is to convert a proto to a JSON-compatible format, but we don't actually want or need that for any reason here. I think it's just used for convenience to call the constructor at the end of the function.

This seems like it would be simpler/less error prone overall. What do you think?

GeneDer replied:
Ok, let me test out how this works 🙂

GeneDer replied:
Ended up implementing this `_proto_to_dict()` to convert protobuf into dictionary as there are too many fields to deal 😅

edoakes commented on python/ray/serve/_private/config.py:
```diff
@@ -48,6 +49,33 @@ def _needs_pickle(deployment_language: DeploymentLanguage, is_cross_language: bo
         return False
 
 
+def _proto_to_dict(proto: Message) -> Dict:
```
Nice this looks pretty simple. Add some basic tests for it?

GeneDer replied:
There are already existing config tests in [test_config.py](https://github.com/ray-project/ray/blob/46aac4db19c82d4fdb2f7f99c83c3bdb3a6ab051/python/ray/serve/tests/unit/test_config.py) so I didn't bother to write new ones. But add some more unit tests for calling `_proto_to_dict ` directly in a bit

shrekris-anyscale commented on python/ray/serve/_private/constants.py:
```diff
@@ -267,3 +267,6 @@
 )
+
+# The default autoscaling policy to use if none is specified.
+DEFAULT_AUTOSCALING_POLICY = "ray.serve.autoscaling_policy:DefaultAutoscalingPolicy"
```
The test can be pretty short:

```python
def test_default_autoscaling_policy_import_path():
    policy = import_attr(DEFAULT_AUTOSCALING_POLICY)

    # Perform a simple assertion to make sure that the policy is as expected
    assert ...
```

This test fails if we move the `DefaultAutoscalingPolicy` but don't update `DEFAULT_AUTOSCALING_POLICY`. That way, if we refactor the codebase in the future, this test makes sure that we also update `DEFAULT_AUTOSCALING_POLICY`'s value.

GeneDer replied:
Gotcha, you don't meant to have the test moving an already imported file and test the error, just add an assertion to make sure the default is importable👍

shrekris-anyscale replied:
Yes that's right, sorry for any confusion.

shrekris-anyscale commented on python/ray/serve/_private/config.py:
```diff
@@ -48,6 +49,33 @@ def _needs_pickle(deployment_language: DeploymentLanguage, is_cross_language: bo
 
 
+def _proto_to_dict(proto: Message) -> Dict:
+    """Recursively convert proto into python dictionary.
```
```suggestion
    """Recursively convert a protobuf into a Python dictionary.
```

shrekris-anyscale commented on python/ray/serve/_private/config.py:
```diff
@@ -48,6 +49,33 @@ def _needs_pickle(deployment_language: DeploymentLanguage, is_cross_language: bo
+def _proto_to_dict(proto: Message) -> Dict:
+    """Recursively convert proto into python dictionary.
+
+    This is needed because MessageToDict handles bytes with extra base64 encoding for
+    constructing json response and is undesirable for our use case.
```
```suggestion
    This is an alternative to protobuf's `MessageToDict`. Unlike 
    `MessageToDict`, this function doesn't add an extra base64
    encoding to bytes when constructing a json response.
```
